### DISCUSSION TITLE
[Question] About Joystick Input (Logging, Joystick Signal Process)

### DISCUSSION THREAD

**Discussion by doltae09** [bitcraze/discussions#643] [Q&A]
Can the command value of a joystick be logged in the Crazyflie firmware?

If Crazyflie flies with only a z-ranger deck attached, how does the joystick input convert to a velocity desired command in the velocity controller?

**Comment by krichardsson** [comment]
I'll try to break down the process to make it easier to find the code.

The gamepad values are not sent directly to the Crazyflie, instead they are converted to setpoints first. A setpoint tells the [controller](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/) what to do and there are a few different flavors of them.
If you look in the [code for the client, that generates the setpoints](https://github.com/bitcraze/crazyflie-clients-python/blob/04393236ed1cc65c4b77953b7687d06a950f2cdd/src/cfclient/ui/main.py#L196-L209), you will find which type of setpoint that is used.

The actual sending of the setpoints is done in [commander in the lib](https://github.com/bitcraze/crazyflie-lib-python/blob/master/cflib/crazyflie/commander.py) if you want to check that out.

In the firmware you will find a bunch of setpoint decoders in [`crtp_commander_generic.c`](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/crtp_commander_generic.c) that converts the setpoint into a generic setpoint format. Finally this is passed on to the controller.

> Can the command value of a joystick be logged in the Crazyflie firmware?

Not directly, but you can for instance log variables in the [`ctrltarget` group](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/logs/#ctrltarget)
